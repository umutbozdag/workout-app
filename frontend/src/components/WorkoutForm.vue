<template>
  <el-form
    :model="workoutFormData"
    :rules="workoutFormRules"
    ref="workoutForm"
    class="workout-form"
  >
    <el-form-item class="workout-form__title" label="Title" prop="title">
      <el-input
        class="workout-form__title-input"
        v-model="workoutFormData.title"
      ></el-input>
    </el-form-item>

    <el-form-item
      class="workout-form__desc"
      label="Description"
      prop="description"
    >
      <el-input
        type="textarea"
        class="workout-form__title-textarea"
        v-model="workoutFormData.description"
      ></el-input>
    </el-form-item>

    <el-button
      type="primary"
      :disabled="isAddWorkoutDisabled"
      @click="handleOnAddWorkout"
      >Add Workout</el-button
    >
  </el-form>
</template>

<script lang="ts">
import Vue from "vue";
import { mapActions, mapState } from "vuex";
import { IWorkout } from "../interfaces";

export default Vue.extend({
  name: "WorkoutForm",
  data() {
    return {
      workoutFormRules: {
        title: [
          {
            required: true,
            message: "Please type a Exercise title",
            trigger: "blur",
          },
        ],
        description: [
          {
            required: true,
            message: "Please type a Exercise description",
            trigger: "blur",
          },
          {
            min: 20,
            max: 240,
            message: "Length should be 20 to 240",
            trigger: "blur",
          },
        ],
      },
    };
  },
  methods: {
    ...mapActions(["addWorkout", "setWorkoutFormData"]),
    handleOnAddWorkout() {
      this.$refs["workoutForm"].validate(async (valid: boolean) => {
        if (valid) {
          try {
            const addedWorkout: IWorkout = await this.addWorkout(
              this.workoutFormData
            );
            this.$notify({
              title: "Success!",
              message: "Workout added successfully",
              type: "success",
            });
            this.$router.push({
              name: "WorkoutDetail",
              params: { workoutId: addedWorkout.workoutId },
            });
            this.setWorkoutFormData({
              title: "",
              description: "",
            });
          } catch (e) {
            this.$notify({
              title: "Error!",
              message: e,
              type: "error",
            });
          }
        } else {
          return false;
        }
      });
    },
  },
  computed: {
    ...mapState(["workoutFormData"]),
    isAddWorkoutDisabled() {
      return (
        !this.workoutFormData.title ||
        !this.workoutFormData.description ||
        this.workoutFormData.description.length < 20 ||
        this.workoutFormData.description.length > 240
      );
    },
  },
});
</script>

<style></style>
