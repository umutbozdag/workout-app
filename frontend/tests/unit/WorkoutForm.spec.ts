import { shallowMount, createLocalVue, mount } from "@vue/test-utils";
import WorkoutForm from "@/components/WorkoutForm.vue";
import Vuex from "vuex";
import { mockWorkoutData } from "./generateMockData";

const localVue = createLocalVue();
localVue.use(Vuex);

describe("WorkoutForm.vue", () => {
  const store = new Vuex.Store({
    state: {
      workoutFormData: {
        title: "",
        description: "",
      },
    },
    actions: {
      addWorkout: jest.fn(),
      setWorkoutFormData: jest.fn(),
    },
  });

  const wrapper = mount(WorkoutForm, {
    store,
    localVue,
  });

  it("Changes title and description correctly", async () => {
    const workoutTitleInputEl = wrapper.find(
      ".workout-form__title .el-input__inner"
    );
    await workoutTitleInputEl.setValue(mockWorkoutData.title);

    const titleInputField: HTMLInputElement =
      workoutTitleInputEl.element as HTMLInputElement;

    expect(titleInputField.value).toBe(mockWorkoutData.title);

    const workoutTitleDesc = wrapper.find(
      ".workout-form__desc .el-textarea__inner"
    );
    await workoutTitleDesc.setValue(mockWorkoutData.description);
    const descInputField: HTMLInputElement =
      workoutTitleDesc.element as HTMLInputElement;

    expect(descInputField.value).toBe(mockWorkoutData.description);
  });

  it("Renders correctly", () => {
    expect(wrapper.element).toMatchSnapshot();
  });
});
