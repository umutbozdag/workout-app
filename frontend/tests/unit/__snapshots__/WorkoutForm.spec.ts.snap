// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WorkoutForm.vue Renders correctly 1`] = `
<form
  class="el-form workout-form"
>
  <div
    class="el-form-item workout-form__title is-required"
  >
    <label
      class="el-form-item__label"
      for="title"
    >
      Title
    </label>
    <div
      class="el-form-item__content"
    >
      <div
        class="workout-form__title-input el-input"
      >
        <!---->
        <input
          autocomplete="off"
          class="el-input__inner"
          type="text"
        />
        <!---->
        <!---->
        <!---->
        <!---->
      </div>
      <transition-stub
        name="el-zoom-in-top"
      >
        <!---->
      </transition-stub>
    </div>
  </div>
   
  <div
    class="el-form-item workout-form__desc is-required"
  >
    <label
      class="el-form-item__label"
      for="description"
    >
      Description
    </label>
    <div
      class="el-form-item__content"
    >
      <div
        class="workout-form__title-textarea el-textarea"
      >
        <textarea
          autocomplete="off"
          class="el-textarea__inner"
          style="min-height: -4px;"
        />
        <!---->
      </div>
      <transition-stub
        name="el-zoom-in-top"
      >
        <!---->
      </transition-stub>
    </div>
  </div>
   
  <button
    class="el-button el-button--primary"
    type="button"
  >
    <!---->
    <!---->
    <span>
      Add Workout
    </span>
  </button>
</form>
`;
